File: paleo_workbench/ui/pages/preview_disk_cache.py
from __future__ import annotations

import hashlib
import json
import logging
import os
import shutil
import uuid
from collections.abc import Callable
from contextlib import AbstractContextManager, nullcontext
from pathlib import Path
from typing import TYPE_CHECKING

import numpy as np

from paleo_workbench.project.models import ResourceItem
from paleo_workbench.ui.pages.preview_cache import safe_file_stat
from paleo_workbench.ui.pages.preview_provider import PreviewResult

if TYPE_CHECKING:
    from geoviz import PreparedPreview, PreviewOptions

logger = logging.getLogger(__name__)

CACHEABLE_RESOURCE_TYPES = frozenset(
    {"horizon", "well_stratification", "well_head"}
)
DIR_NAME = ".preview_cache"


def is_disk_cacheable(asset: object) -> bool:
    return (
        isinstance(asset, ResourceItem)
        and asset.type in CACHEABLE_RESOURCE_TYPES
        and asset.format.strip().lower().lstrip(".") == "dat"
    )


def _options_fingerprint(options: PreviewOptions | None = None) -> str:
    # Runtime import: keep geoviz off cold import of this module.
    from geoviz import PAYLOAD_SCHEMA_VERSION, PreviewOptions as _PreviewOptions

    opts = options or _PreviewOptions.local()
    raw = (
        f"{opts.profile}|{opts.max_curves}|{opts.max_depth_samples}|"
        f"{opts.max_slice_axis}|{opts.max_points}|{opts.surface_grid_size}|"
        f"schema={PAYLOAD_SCHEMA_VERSION}"
    )
    return hashlib.sha256(raw.encode("utf-8")).hexdigest()[:16]


def _entry_key_material(
    asset: ResourceItem,
    options: PreviewOptions | None = None,
    *,
    comparison_crs: str | None = None,
) -> str:
    path = Path(asset.path).resolve()
    st = safe_file_stat(path)
    metadata = asset.parsed_summary or {}
    material = {
        "source_path": str(path),
        "resource_id": asset.id,
        "semantic_type": asset.type,
        "format": asset.format,
        "source_stat": st,
        "checksum": str(asset.checksum or ""),
        "source_crs": str(asset.crs or ""),
        "coordinate_units": str(
            metadata.get("coordinate_units") or metadata.get("units") or ""
        ),
        "comparison_crs": (
            str(comparison_crs)
            if comparison_crs is not None
            else str(metadata.get("comparison_crs") or "")
        ),
        "options": _options_fingerprint(options),
    }
    raw = json.dumps(
        material,
        ensure_ascii=False,
        sort_keys=True,
        separators=(",", ":"),
    )
    return hashlib.sha256(raw.encode("utf-8")).hexdigest()[:32]


def _discard_entry(entry: Path) -> None:
    if entry.exists():
        shutil.rmtree(entry, ignore_errors=True)


class PreviewDiskCache:
    """Project-scoped prepare-result cache under ``.preview_cache/``.

    Complements the in-memory LRU: worker tries disk after a memory miss for
    horizon / well_stratification / well_head DAT only. Corrupt entries are
    deleted on read failure. Store failures never break live preview.
    """

    def __init__(
        self,
        project_root: Path | str | None = None,
        *,
        options: PreviewOptions | None = None,
        comparison_crs: str | None = None,
    ) -> None:
        self.project_root = Path(project_root).resolve() if project_root else None
        self.options = options
        self.comparison_crs = comparison_crs

    def set_options(self, options: PreviewOptions) -> None:
        self.options = options

    def set_project_root(self, project_root: Path | str | None) -> None:
        self.project_root = Path(project_root).resolve() if project_root else None

    def set_comparison_crs(self, comparison_crs: str | None) -> None:
        self.comparison_crs = comparison_crs

    def _entries_dir(self) -> Path | None:
        if self.project_root is None:
            return None
        return self.project_root / DIR_NAME / "entries"

    def try_load(self, asset: ResourceItem) -> PreviewResult | None:
        if self.project_root is None or not is_disk_cacheable(asset):
            return None
        entries = self._entries_dir()
        if entries is None:
            return None
        try:
            key = _entry_key_material(
                asset,
                self.options,
                comparison_crs=self.comparison_crs,
            )
        except Exception:
            logger.warning("preview disk cache key failed for %s", asset.path, exc_info=True)
            return None
        entry = entries / key
        meta_path = entry / "meta.json"
        payload_path = entry / "payload.npz"
        if not meta_path.is_file() or not payload_path.is_file():
            return None
        try:
            from geoviz import decode_prepared_preview

            meta = json.loads(meta_path.read_text(encoding="utf-8"))
            # Re-validate live key matches stored key (mtime/size/options).
            if meta.get("key") != key:
                _discard_entry(entry)
                return None
            with np.load(payload_path, allow_pickle=False) as data:
                arrays = {name: data[name] for name in data.files}
            prepared = decode_prepared_preview(meta["prepared"], arrays)
        except Exception:
            logger.warning(
                "preview disk cache miss (corrupt) for %s at %s",
                asset.path,
                entry,
                exc_info=True,
            )
            _discard_entry(entry)
            return None
        return PreviewResult(
            mode="geoviz",
            title=asset.name,
            path=asset.path,
            format=asset.format,
            status=asset.status,
            type_label=asset.type,
            warning=prepared.warning,
            summary_rows=prepared.summary_rows,
            engine_preview=prepared,
            estimated_bytes=prepared.estimated_bytes,
        )

    def store(
        self,
        asset: ResourceItem,
        result: PreviewResult,
        *,
        commit_guard: Callable[[], AbstractContextManager[bool]] | None = None,
    ) -> None:
        if self.project_root is None or not is_disk_cacheable(asset):
            return
        if result.mode != "geoviz" or result.engine_preview is None:
            return
        try:
            from geoviz import PreparedPreview, encode_prepared_preview
        except ImportError:  # pragma: no cover
            return
        if not isinstance(result.engine_preview, PreparedPreview):
            return
        entries = self._entries_dir()
        if entries is None:
            return
        try:
            key = _entry_key_material(
                asset,
                self.options,
                comparison_crs=self.comparison_crs,
            )
            prepared_meta, arrays = encode_prepared_preview(result.engine_preview)
            entries.mkdir(parents=True, exist_ok=True)
            # Stage under a unique sibling dir, then replace the entry.
            staging = entries / f".tmp-{key}-{uuid.uuid4().hex}"
            staging.mkdir(parents=True, exist_ok=False)
            try:
                meta = {
                    "key": key,
                    "source_path": str(Path(asset.path).resolve()),
                    "semantic_type": asset.type,
                    "prepared": prepared_meta,
                }
                (staging / "meta.json").write_text(
                    json.dumps(meta, ensure_ascii=False, indent=2),
                    encoding="utf-8",
                )
                np.savez_compressed(staging / "payload.npz", **arrays)
                guard = commit_guard() if commit_guard is not None else nullcontext(True)
                with guard as allowed:
                    if not allowed:
                        shutil.rmtree(staging, ignore_errors=True)
                        return
                    entry = entries / key
                    if entry.exists():
                        shutil.rmtree(entry)
                    os.replace(str(staging), str(entry))
            except Exception:
                if staging.exists():
                    shutil.rmtree(staging, ignore_errors=True)
                raise
        except Exception:
            # Preview must still succeed without disk.
            logger.warning(
                "preview disk cache store failed for %s", asset.path, exc_info=True
            )
            return

    def clear(self) -> None:
        if self.project_root is None:
            return
        root = self.project_root / DIR_NAME
        if root.exists():
            shutil.rmtree(root, ignore_errors=True)

File: libs/ui_data_core/src/preview_disk_cache.cpp
// preview_disk_cache.py port — see preview_disk_cache.hpp.

#include "pwb/ui_data_core/preview_disk_cache.hpp"

#include "pwb/domain/sha256.hpp"
#include "pwb/ui_data_core/json_util.hpp"
#include "pwb/ui_data_core/preview_cache.hpp"  // safe_file_stat

#include <algorithm>
#include <cctype>
#include <fstream>
#include <random>

namespace pwb::ui_data_core {

const std::vector<std::string>& disk_cacheable_resource_types() {
    static const std::vector<std::string> types = {
        "horizon", "well_stratification", "well_head"};
    return types;
}

bool is_disk_cacheable(const ResourceItem& asset) {
    const auto& types = disk_cacheable_resource_types();
    if (std::find(types.begin(), types.end(), asset.type) == types.end()) {
        return false;
    }
    // asset.format.strip().lower().lstrip(".") == "dat"
    std::string fmt = asset.format;
    const auto first = fmt.find_first_not_of(" \t\n\r");
    fmt = first == std::string::npos ? "" : fmt.substr(first);
    const auto last = fmt.find_last_not_of(" \t\n\r");
    fmt = last == std::string::npos ? "" : fmt.substr(0, last + 1);
    for (auto& c : fmt) {
        c = static_cast<char>(std::tolower(static_cast<unsigned char>(c)));
    }
    const auto dot = fmt.find_first_not_of('.');
    fmt = dot == std::string::npos ? "" : fmt.substr(dot);
    return fmt == "dat";
}

// ---------------------------------------------------------------------------
// _entry_key_material
// ---------------------------------------------------------------------------

std::string preview_disk_entry_key(
    const ResourceItem& asset, const GeovizPreviewOptions* options,
    const std::optional<std::string>& comparison_crs) {
    const std::filesystem::path resolved =
        std::filesystem::weakly_canonical(std::filesystem::path(asset.path));
    const auto stat = safe_file_stat(resolved);
    const domain::Json& metadata =
        asset.parsed_summary.is_object() ? asset.parsed_summary
                                         : domain::Json::object();

    domain::Json material = domain::Json::object();
    material["source_path"] = resolved.generic_string();
    material["resource_id"] = asset.id;
    material["semantic_type"] = asset.type;
    material["format"] = asset.format;
    if (stat.has_value()) {
        material["source_stat"] = domain::Json::array({stat->first, stat->second});
    } else {
        material["source_stat"] = nullptr;
    }
    material["checksum"] = asset.checksum.value_or("");
    material["source_crs"] = asset.crs.value_or("");
    // str(metadata.get("coordinate_units") or metadata.get("units") or "")
    std::string units;
    if (metadata.is_object() && metadata.contains("coordinate_units") &&
        json_truthy(metadata.at("coordinate_units"))) {
        units = json_str(metadata.at("coordinate_units"));
    } else if (metadata.is_object() && metadata.contains("units") &&
               json_truthy(metadata.at("units"))) {
        units = json_str(metadata.at("units"));
    }
    material["coordinate_units"] = units;
    std::string comparison;
    if (comparison_crs.has_value()) {
        comparison = *comparison_crs;
    } else if (metadata.is_object() && metadata.contains("comparison_crs") &&
               json_truthy(metadata.at("comparison_crs"))) {
        comparison = json_str(metadata.at("comparison_crs"));
    }
    material["comparison_crs"] = comparison;
    const GeovizPreviewOptions default_options;
    material["options"] = options != nullptr ? options->fingerprint()
                                             : default_options.fingerprint();

    // json.dumps(material, ensure_ascii=False, sort_keys=True, separators).
    std::map<std::string, domain::Json> sorted;
    for (auto it = material.begin(); it != material.end(); ++it) {
        sorted.emplace(it.key(), it.value());
    }
    domain::Json sorted_json = domain::Json::object();
    for (const auto& [k, v] : sorted) {
        sorted_json[k] = v;
    }
    const std::string raw = sorted_json.dump();  // ensure_ascii=False
    return domain::Sha256::of_bytes(raw).substr(0, 32);
}

// ---------------------------------------------------------------------------
// PreviewDiskCache
// ---------------------------------------------------------------------------

namespace {

void discard_entry(const std::filesystem::path& entry) {
    std::error_code ec;
    if (std::filesystem::exists(entry, ec)) {
        std::filesystem::remove_all(entry, ec);
    }
}

bool read_file(const std::filesystem::path& path, std::string& out) {
    std::ifstream stream(path, std::ios::binary);
    if (!stream) {
        return false;
    }
    out.assign(std::istreambuf_iterator<char>(stream),
               std::istreambuf_iterator<char>());
    return stream.good() || stream.eof();
}

bool write_file(const std::filesystem::path& path, const std::string& bytes) {
    std::ofstream stream(path, std::ios::binary | std::ios::trunc);
    if (!stream) {
        return false;
    }
    stream.write(bytes.data(), static_cast<std::streamsize>(bytes.size()));
    return stream.good();
}

// uuid4().hex stand-in for staging dir names.
std::string staging_suffix() {
    static std::mt19937_64 rng{std::random_device{}()};
    char buf[33];
    std::snprintf(buf, sizeof(buf), "%016llx%016llx",
                  static_cast<unsigned long long>(rng()),
                  static_cast<unsigned long long>(rng()));
    return buf;
}

}  // namespace

PreviewDiskCache::PreviewDiskCache(
    std::optional<std::filesystem::path> project_root,
    const GeovizPreviewOptions* options,
    std::optional<std::string> comparison_crs)
    : comparison_crs_(std::move(comparison_crs)) {
    if (project_root.has_value()) {
        project_root_ =
            std::filesystem::weakly_canonical(*project_root);
    }
    if (options != nullptr) {
        options_ = *options;
    }
}

void PreviewDiskCache::set_project_root(std::optional<std::filesystem::path> root) {
    if (root.has_value()) {
        project_root_ = std::filesystem::weakly_canonical(*root);
    } else {
        project_root_ = std::nullopt;
    }
}

std::optional<std::filesystem::path> PreviewDiskCache::entries_dir() const {
    if (!project_root_.has_value()) {
        return std::nullopt;
    }
    return *project_root_ / kPreviewDiskCacheDir / "entries";
}

std::optional<PreviewResult> PreviewDiskCache::try_load(
    const ResourceItem& asset) {
    if (!project_root_.has_value() || !is_disk_cacheable(asset)) {
        return std::nullopt;
    }
    const auto entries = entries_dir();
    if (!entries.has_value()) {
        return std::nullopt;
    }
    std::string key;
    try {
        key = preview_disk_entry_key(asset, &options_, comparison_crs_);
    } catch (...) {
        return std::nullopt;
    }
    const std::filesystem::path entry = *entries / key;
    const std::filesystem::path meta_path = entry / "meta.json";
    const std::filesystem::path payload_path = entry / "payload.npz";
    std::error_code ec;
    if (!std::filesystem::is_regular_file(meta_path, ec) || ec ||
        !std::filesystem::is_regular_file(payload_path, ec) || ec) {
        return std::nullopt;
    }
    try {
        std::string meta_text, payload_bytes;
        if (!read_file(meta_path, meta_text) ||
            !read_file(payload_path, payload_bytes)) {
            throw std::runtime_error("read failed");
        }
        const domain::Json meta = domain::Json::parse(meta_text);
        // Re-validate the live key against the stored key.
        if (!meta.is_object() || json_get_string(meta, "key") != key) {
            discard_entry(entry);
            return std::nullopt;
        }
        if (codec_ == nullptr) {
            // geoviz absent → the Python ImportError path: corrupt-class
            // failure, entry discarded, miss.
            throw std::runtime_error("no codec");
        }
        const auto decoded =
            codec_->decode(meta.is_object() && meta.contains("prepared")
                               ? meta.at("prepared")
                               : domain::Json(nullptr),
                           payload_bytes);
        if (!decoded.has_value()) {
            throw std::runtime_error("decode failed");
        }
        PreviewResult result;
        result.mode = preview_mode::kGeoviz;
        result.title = asset.name;
        result.path = asset.path;
        result.format = asset.format;
        result.status = asset.status;
        result.type_label = asset.type;
        result.warning = decoded->warning;
        result.summary_rows = decoded->summary_rows;
        result.engine_preview = decoded->engine_preview;
        result.estimated_bytes = decoded->estimated_bytes;
        return result;
    } catch (...) {
        discard_entry(entry);
        return std::nullopt;
    }
}

void PreviewDiskCache::store(const ResourceItem& asset,
                             const PreviewResult& result,
                             const std::function<bool()>* commit_guard) {
    if (!project_root_.has_value() || !is_disk_cacheable(asset)) {
        return;
    }
    if (result.mode != preview_mode::kGeoviz || result.engine_preview == nullptr) {
        return;
    }
    if (codec_ == nullptr) {
        return;  // geoviz ImportError → silent no-op
    }
    const auto entries = entries_dir();
    if (!entries.has_value()) {
        return;
    }
    try {
        const std::string key =
            preview_disk_entry_key(asset, &options_, comparison_crs_);
        const auto encoded = codec_->encode(result.engine_preview);
        if (!encoded.has_value()) {
            return;
        }
        std::error_code ec;
        std::filesystem::create_directories(*entries, ec);
        if (ec) {
            return;
        }
        // Stage under a unique sibling dir, then replace the entry.
        const std::filesystem::path staging =
            *entries / (".tmp-" + key + "-" + staging_suffix());
        std::filesystem::create_directories(staging, ec);
        if (ec) {
            return;
        }
        try {
            domain::Json meta = domain::Json::object();
            meta["key"] = key;
            meta["source_path"] =
                std::filesystem::weakly_canonical(
                    std::filesystem::path(asset.path))
                    .generic_string();
            meta["semantic_type"] = asset.type;
            meta["prepared"] = encoded->prepared_meta;
            if (!write_file(staging / "meta.json", meta.dump(2)) ||
                !write_file(staging / "payload.npz", encoded->payload_bytes)) {
                throw std::runtime_error("write failed");
            }
            const bool allowed =
                commit_guard == nullptr || (*commit_guard)();
            if (!allowed) {
                discard_entry(staging);
                return;
            }
            const std::filesystem::path entry = *entries / key;
            if (std::filesystem::exists(entry, ec)) {
                std::filesystem::remove_all(entry, ec);
            }
            std::filesystem::rename(staging, entry, ec);
            if (ec) {
                throw std::runtime_error("replace failed");
            }
        } catch (...) {
            discard_entry(staging);
            throw;
        }
    } catch (...) {
        // Preview must still succeed without disk — log-and-continue.
        return;
    }
}

void PreviewDiskCache::clear() {
    if (!project_root_.has_value()) {
        return;
    }
    const std::filesystem::path root = *project_root_ / kPreviewDiskCacheDir;
    std::error_code ec;
    if (std::filesystem::exists(root, ec)) {
        std::filesystem::remove_all(root, ec);
    }
}

}  // namespace pwb::ui_data_core
